### PR TITLE
Fix in a dwarf line information parsing implementation

### PR DESCRIPTION
**Pull request by HoundThe**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**

#16975 
Compilation unit can have multiple line information sequences, radare2 parsed only one because the algorithm was wrong. 

There is one header per compilation unit, and there can be multiple sequences, in the algorithm there wasn't any tracking of read length (there was a loop that was always executed only once, because at the end it appended the length between start to end to the start) 

So I've changed `r_bin_dwarf_parse_opcodes` to return read length so I know if there are still more sequences left. Move the header creation outside the loop, because it's created only once per unit. And I read a sequence every loop.

Tests in #16965  (WIP)

(Compared to previous output shown in the #16975 )
Now:
```
$ r2 cpp_dwarf3
> id

...
Set column to 3
Extended opcode 2: set Address to 0x11ee
Special opcode 6: advance Address by 0 to 11ee and Line by 1 to 2
Set column to 12
Special opcode 173: advance Address by 12 to 11fa and Line by 0 to 2
Set column to 15
Special opcode 201: advance Address by 14 to 1208 and Line by 0 to 2
Advance PC by 3 to 0x120b
Extended opcode 1: End of Sequence
Set column to 11
Extended opcode 2: set Address to 0x120c
Special opcode 7: advance Address by 0 to 120c and Line by 2 to 3
Set column to 21
Special opcode 173: advance Address by 12 to 1218 and Line by 0 to 3
Set column to 22
Special opcode 201: advance Address by 14 to 1226 and Line by 0 to 3
Advance PC by 3 to 0x1229
Extended opcode 1: End of Sequence
Set column to 11
Extended opcode 2: set Address to 0x122a
Special opcode 7: advance Address by 0 to 122a and Line by 2 to 3
Set column to 22
..... // rest, matches output of readelf utility
```

### CODE REVIEW

**Comment by HoundThe** [comment]
The fix changed output of a test
From:
```
0x00091004      488b159d880f.  mov rdx, qword [reloc.vtable_for_std::bad_alloc]  ; [0x1898a8:8]=0
```
To:
```
0x00091004      488b159d880f.  mov rdx, qword [reloc.vtable_for_std::bad_alloc] ; /build/gcc/src/gcc/libstdc++-v3/libsupc++/new:57 ; [0x1898a8:8]=0
```
I guess the latter is the correct one, so I'll change the test

**Comment by HoundThe** [comment]
I've also added some formatting improvement of the line section to make it easier to read

Before:
```
Special opcode 173: advance Address by 12 to 11f7 and Line by 0 to 10
Set column to 23
Special opcode 217: advance Address by 15 to 1206 and Line by 2 to 12
Set column to 1
Special opcode 174: advance Address by 12 to 1212 and Line by 1 to 13
Advance PC by 22 to 0x1228
Extended opcode 1: End of Sequence
DWARF LINE HEADER
  total_length: 361
  version: 3
  header_length: : 43
  mininstlen: 1
  max_ops_per_inst: 0
  is_stmt: 1
  line_base: -5
  line_range: 14
  opcode_base: 13
 op 1 0
 op 2 1
 op 3 1
 op 4 1
 op 5 1
 op 6 0
 op 7 0
 op 8 0
 op 9 1
 op 10 0
 op 11 0
 op 12 1
FILE (mammal.h)
| dir idx 0
| lastmod 0
| filelen 0
FILE (main.cpp)
| dir idx 0
| lastmod 0
| filelen 0
Set column to 3
Extended opcode 2: set Address to 0x12c6
Special opcode 6: advance Address by 0 to 12c6 and Line by 1 to 2
Set column to 12
Special opcode 173: advance Address by 12 to 12d2 and Line by 0 to 2
Set column to 15
Special opcode 201: advance Address by 14 to 12e0 and Line by 0 to 2
Advance PC by 3 to 0x12e3
Extended opcode 1: End of Sequence
```

Now:
```
  Special opcode 173: advance Address by 12 to 1234 and Line by 0 to 7
  Set column to 28
  Special opcode 75: advance Address by 5 to 1239 and Line by 0 to 7
  Advance PC by 2 to 0x123b
  Extended opcode 1: End of Sequence

 Header information:
  Length:                             361
  DWARF Version:                      3
  Header Length:                      43
  Minimum Instruction Length:         1
  Maximum Operations per Instruction: 0
  Initial value of 'is_stmt':         1
  Line Base:                          -5
  Line Range:                         14
  Opcode Base:                        13

 Opcodes:
  Opcode 1 has 0 arg
  Opcode 2 has 1 arg
  Opcode 3 has 1 arg
  Opcode 4 has 1 arg
  Opcode 5 has 1 arg
  Opcode 6 has 0 arg
  Opcode 7 has 0 arg
  Opcode 8 has 0 arg
  Opcode 9 has 1 arg
  Opcode 10 has 0 arg
  Opcode 11 has 0 arg
  Opcode 12 has 1 arg

 The File Name Table:
  Entry Dir     Time      Size       Name
  1     0       0         0          mammal.h
  2     0       0         0          main.cpp

 Line Number Statements:
  Set column to 3
  Extended opcode 2: set Address to 0x12c6
  Special opcode 6: advance Address by 0 to 12c6 and Line by 1 to 2
  Set column to 12
  Special opcode 173: advance Address by 12 to 12d2 and Line by 0 to 2
  Set column to 15
  Special opcode 201: advance Address by 14 to 12e0 and Line by 0 to 2
  Advance PC by 3 to 0x12e3
  Extended opcode 1: End of Sequence

  Set file to 2
  Set column to 3
  Extended opcode 2: set Address to 0x12e4
  Special opcode 8: advance Address by 0 to 12e4 and Line by 3 to 4
```

File: libr/bin/dwarf.c [hunk 1]
```diff
@@ -456,8 +469,9 @@ static const ut8* r_bin_dwarf_parse_ext_opcode(const RBin *a, const ut8 *obuf,
 
 	opcode = *buf++;
 
+	// Maybe add offset to the print later?
```
Comment (XVilka) [inline libr/bin/dwarf.c:472]: Add TODO or FIXME in these cases, so it will be shown in the respective searches or plugins

File: libr/bin/dwarf.c [hunk 2]
```diff
@@ -720,62 +752,77 @@ static const ut8* r_bin_dwarf_parse_opcodes(const RBin *a, const ut8 *obuf,
...
-			unit_length = hdr.unit_length.part2 + 4;
-		} else {
-			unit_length = hdr.unit_length.part1 + 4;
-		}
-		tmplen = R_MIN (tmplen, unit_length);
-		if (tmplen < 1) {
-			r_bin_dwarf_header_fini (&hdr);
-			break;
+
+		ut64 unit_length = (hdr.unit_length.part1 == DWARF_INIT_LEN_64) ? hdr.unit_length.part2 + 4 : hdr.unit_length.part1 + 4;
```
Comment (XVilka) [inline libr/bin/dwarf.c:801]: That line looks a bit too long.

**Pull request merged**